hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/security/client/ClientToAMTokenIdentifier.java

```diff
@@ -18,84 +18,191 @@
+
+import java.io.*;
 
 @Public
 @Evolving
 public class ClientToAMTokenIdentifier extends TokenIdentifier {
 
-  public static final Text KIND_NAME = new Text("YARN_CLIENT_TOKEN");
-
-  private ApplicationAttemptId applicationAttemptId;
-  private Text clientName = new Text();
-
-  // TODO: Add more information in the tokenID such that it is not
-  // transferrable, more secure etc.
-
-  public ClientToAMTokenIdentifier() {
-  }
-
-  public ClientToAMTokenIdentifier(ApplicationAttemptId id, String client) {
-    this();
-    this.applicationAttemptId = id;
-    this.clientName = new Text(client);
-  }
-
-  public ApplicationAttemptId getApplicationAttemptID() {
-    return this.applicationAttemptId;
-  }
-
-  public String getClientName() {
-    return this.clientName.toString();
-  }
-
-  @Override
-  public void write(DataOutput out) throws IOException {
-    out.writeLong(this.applicationAttemptId.getApplicationId()
-      .getClusterTimestamp());
-    out.writeInt(this.applicationAttemptId.getApplicationId().getId());
-    out.writeInt(this.applicationAttemptId.getAttemptId());
-    this.clientName.write(out);
-  }
-
-  @Override
-  public void readFields(DataInput in) throws IOException {
-    this.applicationAttemptId =
-        ApplicationAttemptId.newInstance(
-          ApplicationId.newInstance(in.readLong(), in.readInt()), in.readInt());
-    this.clientName.readFields(in);
-  }
-
-  @Override
-  public Text getKind() {
-    return KIND_NAME;
-  }
-
-  @Override
-  public UserGroupInformation getUser() {
-    if (this.clientName == null) {
-      return null;
+    public static final Text KIND_NAME = new Text("YARN_CLIENT_TOKEN");
+
+    private ClientToAMTokenIdentifier clientToAMTokenIdentifier;
+
+    // TODO: Add more information in the tokenID such that it is not
+    // transferrable, more secure etc.
+
+    public ClientToAMTokenIdentifier() {
+    }
+
+    public ClientToAMTokenIdentifier(ApplicationAttemptId id, String client) {
+        this();
+        clientToAMTokenIdentifier = new ByteArray(id, client);
+    }
+
+    public ApplicationAttemptId getApplicationAttemptID() {
+        return this.clientToAMTokenIdentifier.getApplicationAttemptID();
+    }
+
+    public String getClientName() {
+        return this.clientToAMTokenIdentifier.getClientName();
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        clientToAMTokenIdentifier.write(out);
     }
-    return UserGroupInformation.createRemoteUser(this.clientName.toString());
-  }
 
-  @InterfaceAudience.Private
-  public static class Renewer extends Token.TrivialRenewer {
     @Override
-    protected Text getKind() {
-      return KIND_NAME;
+    public void readFields(DataInput in) throws IOException {
+        byte[] data = IOUtils.readFullyToByteArray(in);
+        try {
+            clientToAMTokenIdentifier = new Proto();
+            clientToAMTokenIdentifier.readFields(new DataInputStream(new ByteArrayInputStream(data)));
```

Comment:
Extending ClientToAMTokenIdentifier for TokenIdentifier implementations seems a bit artificial and leads to override not that usefull methods, with some non necessary overrides + forces to rebuild InputStream on top of the byte[] data to feed readFields method.

Another approach is to define an inner implementation within ClientToAMTokenIdentifier, that could be TokenIdentifier, that defines three methods really needed -> getApplciationAttemptId, getClientName and write.

2 versions of this interface -> ProtoImpl and POJOImpl. Those 2 implems can be constructed directly with a byte[]. If creation of proto fails, create a POJO.

Providing you go to this approach, most other comments become irrelevant.